File: src/main/kotlin/com/zacksimpson/reminders/data/SyncEngine.kt
package com.zacksimpson.reminders.data

import kotlinx.coroutines.flow.first
import kotlinx.serialization.KSerializer
import kotlinx.serialization.json.jsonObject

class SyncException(message: String) : Exception(message)

/**
 * Orchestrates one full sync pass (SYNC_PLAN.md §3 step 4): pull remote, reconcile with
 * local via [SyncLogic]'s merge, push whatever local won, persist the result back to
 * [RemindersRepository]. The impure, network/storage-touching counterpart to
 * [SyncLogic]'s pure (and unit-tested) functions.
 */
class SyncEngine(
    private val remindersRepo: RemindersRepository,
    private val authRepo: AuthRepository,
    private val firestore: FirestoreClient,
) {
    suspend fun sync() {
        val uid = (authRepo.state.first() as? AuthState.SignedIn)?.uid
            ?: throw SyncException("Not signed in")

        val local = remindersRepo.snapshotForSync()
        val remoteLists = fetchCollection(uid, "lists", ReminderList.serializer())
        val remoteTasks = fetchCollection(uid, "tasks", Task.serializer())
        val remoteSettings = firestore.getDocument(uid, "settings", "singleton")
            ?.let { firestoreModelJson.decodeFromJsonElement(Settings.serializer(), it.fields) }

        val listsResult = SyncLogic.mergeCollection(local.lists, remoteLists)
        val tasksResult = SyncLogic.mergeCollection(local.tasks, remoteTasks)
        val settingsResult = SyncLogic.mergeSettings(local.settings, remoteSettings)

        remindersRepo.applySyncedState(listsResult.merged, tasksResult.merged, settingsResult.merged)

        listsResult.toPush.forEach { pushDocument(uid, "lists", it.id, ReminderList.serializer(), it) }
        tasksResult.toPush.forEach { pushDocument(uid, "tasks", it.id, Task.serializer(), it) }
        if (settingsResult.needsPush) {
            pushDocument(uid, "settings", "singleton", Settings.serializer(), settingsResult.merged)
        }

        authRepo.recordSyncSuccess()
    }

    private suspend fun <T> fetchCollection(uid: String, collection: String, serializer: KSerializer<T>): List<T> =
        firestore.listDocuments(uid, collection)
            .map { firestoreModelJson.decodeFromJsonElement(serializer, it.fields) }

    private suspend fun <T> pushDocument(
        uid: String,
        collection: String,
        docId: String,
        serializer: KSerializer<T>,
        value: T,
    ) {
        val fields = firestoreModelJson.encodeToJsonElement(serializer, value).jsonObject
        firestore.setDocument(uid, collection, docId, fields)
    }
}

File: utils/sync/syncEngine.ts
import {
  normalizeList,
  normalizeSettings,
  normalizeTask,
  type ReminderList,
  type Settings,
  type SyncSnapshot,
  type Task,
} from "@/contexts/RemindersContext";
import type { FirestoreClient } from "./firestoreClient";
import type { JsonValue } from "./firestoreValue";
import { mergeCollection, mergeSettings } from "./syncLogic";

export class SyncException extends Error {}

export interface SyncDeps {
  applySyncedState: (
    lists: ReminderList[],
    tasks: Task[],
    settings: Settings
  ) => Promise<void>;
  firestore: FirestoreClient;
  snapshotForSync: () => SyncSnapshot;
  uid: string;
}

async function fetchCollection<T>(
  firestore: FirestoreClient,
  uid: string,
  collection: string,
  normalize: (raw: T) => T
): Promise<T[]> {
  const docs = await firestore.listDocuments(uid, collection);
  return docs.map((doc) => normalize(doc.fields as unknown as T));
}

/**
 * Orchestrates one full sync pass: pull remote, reconcile with local via
 * syncLogic's pure merge, push whatever local won, persist the result back to
 * RemindersContext. Mirrors the native rewrite's SyncEngine.
 */
export async function runSync(deps: SyncDeps): Promise<void> {
  const { firestore, uid, snapshotForSync, applySyncedState } = deps;
  const local = snapshotForSync();

  const remoteLists = await fetchCollection<ReminderList>(
    firestore,
    uid,
    "lists",
    normalizeList
  );
  const remoteTasks = await fetchCollection<Task>(
    firestore,
    uid,
    "tasks",
    normalizeTask
  );
  const remoteSettingsDoc = await firestore.getDocument(
    uid,
    "settings",
    "singleton"
  );
  const remoteSettings = remoteSettingsDoc
    ? normalizeSettings(remoteSettingsDoc.fields as unknown as Settings)
    : null;

  const listsResult = mergeCollection(local.lists, remoteLists);
  const tasksResult = mergeCollection(local.tasks, remoteTasks);
  const settingsResult = mergeSettings(local.settings, remoteSettings);

  await applySyncedState(
    listsResult.merged,
    tasksResult.merged,
    settingsResult.merged
  );

  await Promise.all(
    listsResult.toPush.map((list) =>
      firestore.setDocument(
        uid,
        "lists",
        list.id,
        list as unknown as Record<string, JsonValue>
      )
    )
  );
  await Promise.all(
    tasksResult.toPush.map((task) =>
      firestore.setDocument(
        uid,
        "tasks",
        task.id,
        task as unknown as Record<string, JsonValue>
      )
    )
  );
  if (settingsResult.needsPush) {
    await firestore.setDocument(
      uid,
      "settings",
      "singleton",
      settingsResult.merged as unknown as Record<string, JsonValue>
    );
  }
}
